ros/src/actuation/vehicles/packages/ymc/node/g30esli_interface/g30esli_ros.cpp

```diff
@@ -0,0 +1,329 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "g30esli_ros.h"
+
+G30esliROS::G30esliROS()
+  : alive_(0), reset_command_(false), stop_keep_secs_(1.0), reset_keep_secs_(0.6), disable_reset_secs_(3.0)
+{
+}
+
+G30esliROS::~G30esliROS()
+{
+}
+
+bool G30esliROS::openDevice(const std::string& device)
+{
+  return g30esli_.openDevice(device);
+}
+
+void G30esliROS::updateAutoCommand(const autoware_msgs::VehicleCmd& msg, const bool& engage,
+                                   const double& steering_offset_deg, const double& brake_threshold)
+{
+  Command& cmd = commands_[(int)MODE::AUTO];
+
+  cmd.time = ros::Time::now();
+  cmd.received = true;
+
+  // speed
+  double speed_kmph = msg.ctrl_cmd.linear_velocity * 3.6;  // [m/s] -> [km/h]
+  cmd.command.speed = (engage && !reset_command_) ? speed_kmph : 0.0;
+
+  // steer
+  double steering_angle_deg = msg.ctrl_cmd.steering_angle / M_PI * 180.0;  // [rad] -> [deg]
+  cmd.command.steer = -(steering_angle_deg + steering_offset_deg);
+
+  // mode
+  cmd.command.mode = engage ? G30ESLI_MODE_AUTO : G30ESLI_MODE_MANUAL;
+
+  // brake
+  if (msg.emergency == 1)
+  {
+    cmd.command.brake = G30ESLI_BRAKE_SEMIEMG;
+  }
+  else if (cmd.command.speed < brake_threshold)
+  {
+    cmd.command.brake = G30ESLI_BRAKE_SMOOTH;
+  }
+  else
+  {
+    cmd.command.brake = G30ESLI_BRAKE_NONE;
+  }
+
+  // shift
+  if (msg.gear == 1)
+  {
+    cmd.command.shift = G30ESLI_SHIFT_DRIVE;
+  }
+  else if (msg.gear == 2)
+  {
+    cmd.command.shift = G30ESLI_SHIFT_REVERSE;
+  }
+  else if (msg.gear == 4)
+  {
+    cmd.command.shift = G30ESLI_SHIFT_NEUTRAL;
+  }
+
+  // flasher
+  if (msg.lamp_cmd.l == 0 && msg.lamp_cmd.r == 0)
+  {
+    cmd.command.flasher = G30ESLI_FLASHER_CLEAR;
+  }
+  else if (msg.lamp_cmd.l == 1 && msg.lamp_cmd.r == 0)
+  {
+    cmd.command.flasher = G30ESLI_FLASHER_LEFT;
+  }
+  else if (msg.lamp_cmd.l == 0 && msg.lamp_cmd.r == 1)
+  {
+    cmd.command.flasher = G30ESLI_FLASHER_RIGHT;
+  }
+  else if (msg.lamp_cmd.l == 1 && msg.lamp_cmd.r == 1)
+  {
+    cmd.command.flasher = G30ESLI_FLASHER_HAZARD;
+  }
+}
+
+void G30esliROS::updateJoystickCommand(const ds4_msgs::DS4& msg, const bool& engage, const double& steering_offset_deg)
+{
+  Command& cmd = commands_[(int)MODE::JOYSTICK];
+
+  cmd.time = ros::Time::now();
+  cmd.received = true;
+
+  // check connection
+  if (!msg.connected)
+  {
+    cmd.command.speed = 0.0;
+    cmd.command.brake = G30ESLI_BRAKE_SEMIEMG;
+    return;
+  }
+
+  // speed
+  cmd.command.speed = msg.cross ? 16.0 * msg.r2 + 3.0 : 0.0;
```

Comment:
https://github.com/autowarefoundation/autoware/blob/d720d23de83de708d9a59d6d8c17986003aa8c49/ros/src/actuation/vehicles/packages/ymc/node/g30esli_interface/g30esli_ros.cpp#L43
The same idea as above, but for `Joystick` we need to assume `engage` and `reset_command_`.
```suggestion
  bool send_speed = msg.cross && engage && !reset_command_;
  cmd.command.speed = send_speed ? 16.0 * msg.r2 + 3.0 : 0.0;
```

Reply:
Good point. Thanks.